ProcessLib/PhaseField/PhaseFieldFEM.h

```diff
@@ -0,0 +1,609 @@
+                    .noalias() = N;
+
+            auto const rho_sr = _process_data.solid_density(t, x_position)[0];
+            auto const& b = _process_data.specific_body_force;
+            local_rhs.template segment<displacement_size>(displacement_index)
+                .noalias() -=
+                (B.transpose() * sigma_real - N_u.transpose() * rho_sr * b) * w;
+
+            //
+            // displacement equation, phasefield part
+            //
+            Kud.noalias() += B.transpose() * 2 * d_ip * sigma_tensile * N * w;
+
+            double const d_dot_ip = N.dot(d_dot);
+
+            if (history_variable_prev < strain_energy_tensile)
+            {
+                history_variable = strain_energy_tensile;
+                Kdu.noalias() = Kud.transpose();
+            }
+            else
+            {
+                history_variable = history_variable_prev;
+            }
+
+            //
+            // phasefield equation, phasefield part.
+            //
+            Kdd.noalias() += (Kdd_1 + N.transpose() * 2 * history_variable * N +
+                              N.transpose() * 0.5 * gc / ls * N) *
+                             w;
+            double const M =
+                _process_data.kinetic_coefficient(t, x_position)[0];
+            local_rhs.template segment<phasefield_size>(phasefield_index)
+                .noalias() -= (N.transpose() * d_dot_ip / M + Kdd_1 * d +
+                               N.transpose() * d_ip * 2 * history_variable -
+                               N.transpose() * 0.5 * gc / ls * (1 - d_ip)) *
+                              w;
+
+            Ddd.noalias() += N.transpose() / M * N * w;
+        }
+        // displacement equation, phasefield part
+        local_Jac.template block<displacement_size, phasefield_size>(
+                     displacement_index, phasefield_index)
+            .noalias() += Kud;
+
+        // phasefield equation, phasefield part.
+        local_Jac.template block<phasefield_size, phasefield_size>(
+                     phasefield_index, phasefield_index)
+            .noalias() += Kdd + Ddd / dt;
+
+        // phasefield equation, displacement part.
+        local_Jac.template block<phasefield_size, displacement_size>(
+                     phasefield_index, displacement_index)
+            .noalias() += Kdu;
+
+        // Eigen::EigenSolver<JacobianMatrix> eigensolver(local_Jac);
+        // std::cout << "eigenvalues" << eigensolver.eigenvalues() << "\n";
+    }
+
+    void preTimestepConcrete(std::vector<double> const& /*local_x*/,
+                             double const /*t*/,
+                             double const /*delta_t*/) override
+    {
+        unsigned const n_integration_points =
+            _integration_method.getNumberOfPoints();
+
+        for (unsigned ip = 0; ip < n_integration_points; ip++)
+        {
+            _ip_data[ip].pushBackState();
+            if (_ip_data[ip].history_variable_prev <
```

Comment:
The if is not needed. Push back probably always has to happen. The if is done above and determines which value is in history_variable (either history_variable_prev or the new tensile SED).